jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java

```diff
@@ -0,0 +1,259 @@
+ * by their text labels using entered value
+ */
+public class MenuSearchField extends Composite<TextField>
+        implements HasHelper, HasPrefix, HasSuffix, HasValueChangeMode, HasAutocapitalize,
+        HasThemeVariant<MenuSearchFieldVariant>, HasTitle, HasClearButton, HasAutofocus, HasEnabled, HasLabel, HasSize,
+        HasStyle, HasTooltip, InputNotifier, KeyNotifier, HasAriaLabel, Focusable<MenuSearchField>, HasPlaceholder,
+        HasValueAndElement<ComponentValueChangeEvent<MenuSearchField, String>, String> {
+
+    protected static final String SEARCH_FIELD_STYLE_NAME = "jmix-menu-search-field";
+
+    protected MenuItemProvider<?> menuItemProvider;
+    protected MenuItemsTransformer<?> itemsTransformer;
+
+    @Override
+    protected TextField initContent() {
+        TextField field = super.initContent();
+        field.addValueChangeListener(this::onValueChange);
+        field.addClassName(SEARCH_FIELD_STYLE_NAME);
+
+        return field;
+    }
+
+    protected void onValueChange(HasValue.ValueChangeEvent<String> event) {
+        if (menuItemProvider != null) {
+            itemsTransformer.setSearchString(event.getValue());
+            menuItemProvider.load();
+        }
+    }
+
+    /**
+     * @return menu item provider used to filter menu items
+     */
+    @Nullable
+    public MenuItemProvider<?> getMenuItemProvider() {
+        return menuItemProvider;
+    }
+
+    /**
+     * Sets menu item provider which will be used to filter menu items
+     *
+     * @param menuItemProvider menu item provider to set
+     * @param <T>              menu item type
+     */
+    public <T extends MenuItem> void setMenuItemProvider(@Nullable MenuItemProvider<T> menuItemProvider) {
+        this.menuItemProvider = menuItemProvider;
+        if (menuItemProvider != null) {
+            MenuItemsTransformer<T> itemsTransformer = new MenuItemsTransformer<>();
+            menuItemProvider.addMenuItemsTransformer(itemsTransformer);
+            this.itemsTransformer = itemsTransformer;
+        } else {
+            this.itemsTransformer = null;
+        }
+    }
+
+    @Override
+    public ValueChangeMode getValueChangeMode() {
+        return getContent().getValueChangeMode();
+    }
+
+    @Override
+    public void setValueChangeMode(ValueChangeMode valueChangeMode) {
+        getContent().setValueChangeMode(valueChangeMode);
+    }
+
+    @Override
+    public void setValueChangeTimeout(int valueChangeTimeout) {
+        getContent().setValueChangeTimeout(valueChangeTimeout);
+    }
+
+    @Override
+    public int getValueChangeTimeout() {
+        return getContent().getValueChangeTimeout();
+    }
+
+    @Override
+    public void setValue(String value) {
+        getContent().setValue(value);
+    }
+
+    @Override
+    public String getValue() {
+        return getContent().getValue();
+    }
+
+    @Override
+    public Registration addValueChangeListener(
```

Comment:
I have doubt that we need `ValueChangeListener`, at least the one that works simultaneously with the default one. For example, for actions when `ActionPerformedEvent` listener is added don't execute the default logic.